internal/plans/deferring/deferred.go

```diff
@@ -435,3 +436,9 @@ func (d *Deferred) ReportModuleExpansionDeferred(addr addrs.PartialExpandedModul
 	}
 	d.partialExpandedModulesDeferred.Add(addr)
 }
+
+// UnexpectedProviderDeferralDiagnostic is a diagnostic that indicates that a
+// provider was deferred although deferrals were not allowed.
+func UnexpectedProviderDeferralDiagnostic(addrs addrs.AbsResourceInstance) tfdiags.Diagnostic {
+	return tfdiags.Sourceless(tfdiags.Error, "Provider deferred changes when Terraform did not allow deferrals", fmt.Sprintf("The provider signaled a deferred action for %q, but in this context deferrals are disabled. This is a bug in the provider, please file an issue.", addrs.String()))
```

Comment:
Nit: I think your team has specific wording they may want, but I hope this suggestion at least clears up where the issue should be filed (hint: not the core repo)

```suggestion
	return tfdiags.Sourceless(tfdiags.Error, "Provider deferred changes when Terraform did not allow deferrals", fmt.Sprintf("The provider signaled a deferred action for %q, but in this context deferrals are disabled. This is a bug in the provider, please file an issue with the provider developers.", addrs.String()))
```